File: 2904-shortest-and-lexicographically-smallest-beautiful-string/solution.py
class Solution:
    def shortestBeautifulString(self, s: str, k: int) -> str:
        # Collect 0-based indices of all '1's
        ones = [i for i, ch in enumerate(s) if ch == '1']
        
        # If total '1's is less than k, no beautiful substring is possible
        if len(ones) < k:
            return ""
        
        ans = ""
        
        # Check every window of k '1's
        for i in range(len(ones) - k + 1):
            start = ones[i]
            end = ones[i + k - 1]
            sub = s[start : end + 1]
            
            # Select the shortest string; break ties lexicographically
            if not ans or len(sub) < len(ans) or (len(sub) == len(ans) and sub < ans):
                ans = sub
                
        return ans

File: 2904-shortest-and-lexicographically-smallest-beautiful-string/solution.cpp
#include <string>
#include <vector>
#include <algorithm>

class Solution {
public:
    std::string shortestBeautifulString(std::string s, int k) {
        std::vector<int> ones;
        for (int i = 0; i < s.length(); ++i) {
            if (s[i] == '1') {
                ones.push_back(i);
            }
        }

        // If there are fewer than k '1's, no beautiful substring exists
        if (ones.size() < k) {
            return "";
        }

        std::string ans = "";

        // Iterate through all windows of k '1's
        for (size_t i = 0; i + k - 1 < ones.size(); ++i) {
            int start = ones[i];
            int end = ones[i + k - 1];
            std::string sub = s.substr(start, end - start + 1);

            // Update ans if sub is shorter, or same length but lexicographically smaller
            if (ans.empty() || sub.length() < ans.length() || 
               (sub.length() == ans.length() && sub < ans)) {
                ans = sub;
            }
        }

        return ans;
    }
};